Date: Thu, 15 Sep 2016 20:23:51 -0400
From: Felix Janda <felix.janda@...teo.de>
To: musl@...ts.openwall.com
Subject: [PATCH] add pthread_setname_np

the thread name can be displayed using gdb's "info threads".
---
 include/pthread.h               |  1 +
 src/thread/pthread_setname_np.c | 25 +++++++++++++++++++++++++
 2 files changed, 26 insertions(+)
 create mode 100644 src/thread/pthread_setname_np.c

diff --git a/include/pthread.h b/include/pthread.h
index 3d2e0c4..94ef919 100644
--- a/include/pthread.h
+++ b/include/pthread.h
@@ -214,6 +214,7 @@ struct cpu_set_t;
 int pthread_getaffinity_np(pthread_t, size_t, struct cpu_set_t *);
 int pthread_setaffinity_np(pthread_t, size_t, const struct cpu_set_t *);
 int pthread_getattr_np(pthread_t, pthread_attr_t *);
+int pthread_setname_np(pthread_t, const char *);
 int pthread_tryjoin_np(pthread_t, void **);
 int pthread_timedjoin_np(pthread_t, void **, const struct timespec *);
 #endif
diff --git a/src/thread/pthread_setname_np.c b/src/thread/pthread_setname_np.c
new file mode 100644
index 0000000..e328c0d
--- /dev/null
+++ b/src/thread/pthread_setname_np.c
@@ -0,0 +1,25 @@
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/prctl.h>
+
+#include "pthread_impl.h"
+
+int pthread_setname_np(pthread_t thread, const char *name)
+{
+	int fd, cs, status = 0;
+	char f[sizeof "/proc/self/task//comm" + 3*sizeof(int)];
+	size_t len;
+
+	if ((len = strnlen(name, 16)) > 15) return ERANGE;
+
+	if (thread == pthread_self())
+		return prctl(PR_SET_NAME, (unsigned long)name, 0, 0, 0) ? errno : 0;
+
+	snprintf(f, sizeof f, "/proc/self/task/%d/comm", thread->tid);
+	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cs);
+	if ((fd = open(f, O_WRONLY)) < 0 || write(fd, name, len) < 0) status = errno;
+	if (fd >= 0) close(fd);
+	pthread_setcancelstate(cs, 0);
+	return status;
+}
